### PR TITLE
Draft PR for twXXInput functions

### PR DESCRIPTION
**Pull request by DavZim**
Lets discuss and continue our discussion for the inputs here...

### CODE REVIEW

**Comment by kylebutts** [comment]
This is really awesome man. Should I merge what you have right now and then more later? Or want to wait?

**Comment by DavZim** [comment]
Let's wait and discuss before we merge

**Comment by DavZim** [comment]
Still missing:

 - [x] dateInput
 - [x] dateRangeInput
 - [x] sliderInput
 - [x] selectizeInput (partially covered by selectInput with selectize enabled)
 - [x] varSelectizeInput (partially covered by varSelectInput with selectize enabled)

**Comment by DavZim** [comment]
@kylebutts, all inputs should be done now. Have a look around and feel free to merge if you feel like the code and the examples are good!

**Comment by kylebutts** [comment]
@DavZim Awesome man! I'm pretty swamped the next two days, but I'll look ASAP

**Comment by DavZim** [comment]
No worries. I feel this is now large enough that it requires some time to review :)
Let me know if you have any questions/comments.

Also, the latest commit ([0654fc9](https://github.com/kylebutts/shiny.tailwind/pull/7/commits/0654fc985b757cea81d5c32407576b88179e0ac8)) adds tabs which enable simple dashboards etc. effectively closing #8.

I am not quite happy with the example in `inst/examples/06-sidebar-dashboard` as well as the examples in the documentation.
Feel free to improve the examples and make them nicer looking.

File: R/twSelectInput.R [hunk 1]
```diff
@@ -43,7 +43,7 @@
 #' )
 #' }
 twSelectInput <- function(inputId, label, choices, selected = NULL,
-                          multiple = FALSE, selectize = TRUE, width = NULL, size = NULL,
```
Comment (DavZim) [inline R/twSelectInput.R]: shiny::selectInput uses selectize = TRUE. Do we want to deviate from this here?

**Comment by DavZim** [comment]
We could think of adding `use_flowbite()` as well?! what do you think.
As far as I see it it just requries a head entry for the css and another CDN link to the js file. [source](https://flowbite.com/docs/getting-started/introduction/#include-via-cdn)

File: R/twSelectInput.R (hunk 1 above)
Comment (kylebutts) [reply R/twSelectInput.R]: Ohh hmm, my comment yesterday must have not worked. I was having a lot of problems with selective importing its custom files. I saw you trying to fix it, but I'm not sure how it is still loading.

Could you try loading example 4 with `selective = TRUE` and tell me if it happens to you?

**Comment by kylebutts** [comment]
I do think flowbite could be cool to add in as well. I'm going to take a thorough look through stuff, I promise! Just on a work trip all week, so quite behind on open source stuff right now. If you feel comfortable with everything, we can merge for now and then I'll look through next week and push some fixes if anything seems off?

**Comment by DavZim** [comment]
No worries, I think that is expected when working on open source stuff.
When we are looking at flowbite, we might also want to look at daisyUI (which also has a [CDN](https://daisyui.com/docs/cdn/) version). Note that both flowbite and daisyUI recommend the npm install for deployment. It would be neat to have tailwindcss-CLI also work with them, but I haven't checked this and its out of our control.

Ill check example 4 later when I find the time.

File: R/twSelectInput.R (hunk 1 above)
Comment (DavZim) [reply R/twSelectInput.R]: I have tried the following example and it works as expected from my side:
```R
library(shiny),
library(shiny.tailwind)

ui <- div(
  class = "grid grid-cols-2 gap-8",
  use_tailwind(),
  
  div(
    class = "border rounded-md bg-stone-200 shadow-md p-4",
    div(HTML("<code>shiny</code> inputs"), class = "font-bold text-4xl"),
    selectInput("id1", "Selectize", letters[1:3], multiple = TRUE, selectize = TRUE),
    selectInput("id2", "Select", letters[1:3], multiple = TRUE, selectize = FALSE)
  ),
  div(
    class = "border rounded-md bg-stone-200 shadow-md p-4",
    div(HTML("<code>shiny.tailwind</code> inputs"), class = "font-bold text-4xl"),
    twSelectInput("id3", "Selectize", letters[1:3], multiple = TRUE, selectize = TRUE),
    twSelectInput("id4", "Select", letters[1:3], multiple = TRUE, selectize = FALSE)
  )
)

shinyApp(ui, function(input, output) {})
```
image

**Comment by DavZim** [comment]
I think the PR is now in a state that we can merge it and potentially publish to CRAN.
What is your opinion? We can also have a quick call if you prefer and I will walk you through the different changes.
Lastly, I have checked example 4 and it works with selectize = TRUE.

**Comment by DavZim** [comment]
One thing that came to my mind, we could rename the tabset functions to `twMenuItem` and `twTabItem` to align the naming to shiny

**Comment by kylebutts** [comment]
Still getting the `selectize = TRUE` problem, hmm. I tried in Firefox and Safari and got it in both..

Otherwise, this is very nice. The examples are really nice and thorough, thanks for really filling out the package. Should be good to merge and then publish on CRAN. If you want to change the tablet functions, go ahead and then I'll merge.

**Comment by DavZim** [comment]
I just checked again and shiny uses a different naming system, which is also different from shinydashboard... But I think our naming for tabs is good enough and makes it clear that it can also be styled to be something different than a sidebar tab layout.

I have also tested the selectize again on 2 different windows machines and it works on both...

From my perspective we are good to merge and send off to CRAN.

**Comment by DavZim** [comment]
Hi Kyle, do you have a general timeline for this PR, a merge and a possible submission to CRAN?

**Comment by kylebutts** [comment]
I'll submit on August 6th it looks like! https://cran.r-project.org/submit.html

**Comment by kylebutts** [comment]
@DavZim I'm getting a weird error that only shows up when I submit to CRAN. Checked that file and nothing seems wrong. It's only a bug on windows which I can't easily test.

```text
* using log directory 'd:/RCompile/CRANincoming/R-devel/shiny.tailwind.Rcheck'
* using R Under development (unstable) (2022-08-11 r82713 ucrt)
* using platform: x86_64-w64-mingw32 (64-bit)
* using session charset: UTF-8
* checking for file 'shiny.tailwind/DESCRIPTION' ... OK
* checking extension type ... Package
* this is package 'shiny.tailwind' version '0.2.0'
* package encoding: UTF-8
* checking CRAN incoming feasibility ... NOTE
Maintainer: 'Kyle Butts <kyle.butts@colorado.edu>'

New submission

Possibly misspelled words in DESCRIPTION:
  TailwindCSS (3:8, 15:21)
  css (16:21)
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking whether package 'shiny.tailwind' can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking for future file timestamps ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking loading without being on the library search path ... OK
* checking use of S3 registration ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... [8s] OK
* checking Rd files ... [1s] OK
* checking Rd metadata ... OK
* checking Rd line widths ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking examples ... [4s] OK
* checking PDF version of manual ... OK
* checking HTML version of manual ... NOTE
Found the following HTML validation problems:
twTextInput.html:73:50: Warning: <a> illegal characters found in URI

* checking for detritus in the temp directory ... OK
* DONE
Status: 2 NOTEs
```

**Comment by DavZim** [comment]
I think there is an extra `]` in the documentation of `twTextInput()` (ie `. see also [MDN Input Types](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#input_types])` should be `. see also [MDN Input Types](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#input_types)` 

Also maybe the css in the DESCRIPTION file does not throw a warning when used as CSS? TailwindCSS is clearly spelt correctly.

**Comment by DavZim** [comment]
I think there is an extra `]` in the documentation of `twTextInput()` (ie `. see also [MDN Input Types](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#input_types])` should be `. see also [MDN Input Types](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#input_types)` 

Also maybe the css in the DESCRIPTION file does not throw a warning when used as CSS? TailwindCSS is clearly spelt correctly.

**Comment by DavZim** [comment]
Did that help with the CRAN comments?

**Comment by kylebutts** [comment]
Edit: It was the `@` sign in the email. 

Okay, I think everything should be good. It's submitted. CRAN checks can be such a pain in the neck

**Comment by kylebutts** [comment]
Nope! Still getting `twTextInput.html:73:50: Warning: <a> illegal characters found in URI` as an error after fixing that. It doesn't show up locally, so I have no clue how to troubleshoot this

**Comment by DavZim** [comment]
Let me know if there are any other problems with the CRAN submission.

**Comment by DavZim** [comment]
Any luck with the submission?

**Comment by kylebutts** [comment]
It looks like it passed all the tests, so I'm just waiting on approval (CRAN does a manual check for first time submissions).

**Comment by kylebutts** [comment]
@DavZim Here is what one of the CRAN maintainers said: 

```
Thanks,

Please always write package names, software names and API (application
programming interface) names in single quotes in title and description.
e.g: --> 'TailwindCSS '
Please note that package names are case sensitive.

Please use only undirected quotation marks in the description text.
e.g. `@apply` --> '@apply'

Please add \value to .Rd files regarding exported methods and explain
the functions results in the documentation. Please write about the
structure of the output (class) and also what the output means.
(If a function does not return a value, please document that too, e.g.
\value{No return value, called for side effects} or similar)
issing Rd-tags in up to 15 .Rd files, e.g.:
      install_tailwindcss_cli.Rd: \value
      twCheckboxGroupInput.Rd: \value
      twCheckboxInput.Rd: \value
      twDateInput.Rd: \value
      twDateRangeInput.Rd: \value
      twFileInput.Rd: \value
      ...

Please ensure that your functions do not write by default or in your
examples/vignettes/tests in the user's home filespace (including the
package directory and getwd()). This is not allowed by CRAN policies.
Please omit any default path in writing functions.
In your examples/vignettes/tests you can write to tempdir().

Please fix and resubmit.
```

**Comment by DavZim** [comment]
Sounds reasonable, do you want me to look into it?

**Comment by kylebutts** [comment]
If you could, that'd be great. I think almost all of it is pretty simple. I'd copy the `\value` from `htmltools`: https://rstudio.github.io/htmltools/reference/builder.html or from `shiny`: https://shiny.rstudio.com/reference/shiny/1.7.0/selectInput.html

I think the last part is the more difficult one. We should make the install_tailwind and the compile css functions error by default with a message, e.g. "You must specify the tailwind destination." (maybe with a hint for a global installation?) and "You must specify where you want the resulting css file"?

**Pull request merged**